tests/utils/sharing.py

```diff
@@ -0,0 +1,41 @@
+import json
+from typing import Any, Set
+
+
+class IdHashable:
+    obj: Any
+
+    def __init__(self, obj: Any) -> None:
+        self.obj = obj
+
+    def __hash__(self):
+        return id(self.obj)
+
+    def __eq__(self, other: object) -> bool:
+        return (type(self), hash(self)) == (type(other), hash(other))
+
+    def __repr__(self):
+        return json.dumps(self.obj)
+
+
+def collect_mutable_objects(a: Any) -> Set[IdHashable]:
+    ret: set[IdHashable] = set()
+
+    def _register(obj: Any):
+        if isinstance(obj, (dict, list, tuple)):
+            ret.add(IdHashable(obj))
+
+    def _rec(obj: Any):
+        _register(obj)
+        if isinstance(obj, dict):
+            list(map(_rec, obj.values()))
+        elif isinstance(obj, (list, tuple)):
+            list(map(_rec, obj))
+
+    _rec(a)
+
+    return ret
+
+
+def collect_shared_mutable_objects(a: Any, b: Any) -> Set[IdHashable]:
```

Comment:
I need a wrapper that combines the value and its id to ease debug of test failures:
Without the value, I will only see the set of overlapping ids (ints) which doesn't help me much.
Moreover, the ids will change on the second run of the tests.

Reply:
By knowing object id you could easily get object value, if you want this for debugging purposes:

```python
import ctypes
import json
from typing import Any, Set


def collect_all_ids(obj: Any) -> Set[int]:
    ret = {id(obj)}
    if isinstance(obj, dict):
        for v in obj.values():
            ret.update(collect_all_ids(v))
    elif isinstance(obj, (list, tuple)):
        for v in obj:
            ret.update(collect_all_ids(v))
    return ret


def get_object_value(object_id: int):
    return json.dumps(ctypes.cast(object_id, ctypes.py_object).value)


def get_shared_objects_values(a, b) -> Set[str]:
    return set(
        get_object_value(value)
        for value in (collect_all_ids(a) & collect_all_ids(b))
    )
```

But I don't know, that 25 lines vs 41, not a much of difference

If you still wanna stick to your approach, I'm ok with it

Reply:
`get_object_value` isn't going to work reliably. The original value might have been garbage collected, so I may end up with a segfault error or retrieve a different object _(in case if the id was reused)_.